src/ast/transform/MaterializeAggregationQueries.cpp

```diff
@@ -45,197 +46,305 @@
-            // create the body of the new materialised rule
-            for (const auto& cur : agg.getBodyLiterals()) {
-                aggClause->addToBody(souffle::clone(cur));
-            }
-            // find stuff for which we need a grounding
-            for (const auto& argPair : analysis::getGroundedTerms(translationUnit, *aggClause)) {
-                const auto* variable = dynamic_cast<const ast::Variable*>(argPair.first);
-                bool variableIsGrounded = argPair.second;
-                // if it's not even a variable type or the term is grounded
-                // then skip it
-                if (variable == nullptr || variableIsGrounded) {
-                    continue;
-                }
+std::set<std::string> MaterializeAggregationQueriesTransformer::distinguishHeadArguments(
+        const TranslationUnit& tu, const Clause& clause, const Aggregator& aggregate) {
+    /**
+     * The head atom should contain immediate local and injected variables.
+     * No witnesses! They have already been transformed away.
+     * This means that we exclude any inner aggregate local variables. But
+     * we do NOT exclude inner aggregate injected variables!! It's important
+     * that the injected variable ends up in this head so that we do not obfuscate
+     * the injected variable's relationship to the outer scope.
+     * It does not affect the aggregate value because adding an extra column
+     * for the injected variable, where that column will only have one value at a time,
+     * will essentially replicate the aggregate body relation for as many possible
+     * values of the injected variable that there are. The fact that the injected variable
+     * will take one value at a time is key.
+     **/
+    std::set<std::string> headArguments;
+    // find local variables of this aggregate and add them
+    for (const auto& localVarName : analysis::getLocalVariables(tu, clause, aggregate)) {
+        headArguments.insert(localVarName);
+    }
+    // find local variables of inner aggregate and remove them
+    visitDepthFirst(aggregate, [&](const Aggregator& innerAggregate) {
+        if (aggregate == innerAggregate) {
+            return;
+        }
+        for (const auto& innerLocalVariableName : analysis::getLocalVariables(tu, clause, innerAggregate)) {
+            headArguments.erase(innerLocalVariableName);
+        }
+    });
+    // find injected variables of this aggregate and add them
+    for (const auto& injectedVarName : analysis::getInjectedVariables(tu, clause, aggregate)) {
+        headArguments.insert(injectedVarName);
+    }
+    return headArguments;
+}
 
-                for (const auto& lit : clause.getBodyLiterals()) {
-                    const auto* atom = dynamic_cast<const Atom*>(lit);
-                    if (atom == nullptr) {
-                        continue;  // it's not an atom so it can't help ground anything
+// TODO: Deal with recursive parameters with an assert statement.
```

Comment:
Add ownership or issue number to the TODO

Reply:
what is ownership?

Reply:
I suppose I should be doing it but I really feel like I've run out of time at this point.

Reply:
Ownership of the TODO - some record of who is responsible or what the issue is.
```
// TODO (rdowavic): Deal with ...
```
or
```
// TODO (#<issuenum>): Deal with ...
```

Reply:
Weird. The changes weren't showing up, but commenting on it made them show up immediately. All good now.